torch_xla/csrc/tensor.cpp

```diff
@@ -899,15 +900,21 @@ std::vector<xla::ComputationClient::DataPtr> XLATensor::GatherTensorsXlaData(
     const std::vector<XLATensor>& tensors, absl::Span<const size_t> indices,
     absl::Span<const xla::ComputationClient::DataPtr> tensors_data) {
   std::vector<xla::ComputationClient::DataPtr> result_tensors_data;
+  std::unordered_map<xla::int64, xla::ComputationClient::DataPtr>
+      tensor_data_map;
   size_t indices_index = 0;
   for (size_t i = 0; i < tensors.size(); ++i) {
-    if (indices_index < indices.size() && i == indices[indices_index]) {
+    xla::int64 tensor_id = tensors[i].GetUniqueId();
```

Comment:
Thanks for catching this!
I'd probably do it like this, as there is one lookup less:
```diff
--- a/torch_xla/csrc/tensor.cpp
+++ b/torch_xla/csrc/tensor.cpp
@@ -899,15 +899,19 @@ std::vector<xla::ComputationClient::DataPtr> XLATensor::GatherTensorsXlaData(
     const std::vector<XLATensor>& tensors, absl::Span<const size_t> indices,
     absl::Span<const xla::ComputationClient::DataPtr> tensors_data) {
   std::vector<xla::ComputationClient::DataPtr> result_tensors_data;
+  std::unordered_map<xla::int64, size_t> uid_index_map;
   size_t indices_index = 0;
   for (size_t i = 0; i < tensors.size(); ++i) {
-    if (indices_index < indices.size() && i == indices[indices_index]) {
+    xla::int64 tensor_id = tensors[i].GetUniqueId();
+    auto it = uid_index_map.find(tensor_id);
+    if (it != uid_index_map.end()) {
+      result_tensors_data.push_back(result_tensors_data[it->second]);
+    } else if (indices_index < indices.size() && i == indices[indices_index]) {
       // If we are at the current index (it means that the tensor at index
       // 'i' had an IR node to sync, use the XLA data held within the Async
       // object.
-      if (!tensors[i].CurrentTensorData()) {
-        result_tensors_data.push_back(tensors_data[indices_index]);
-      }
+      uid_index_map.emplace(tensor_id, result_tensors_data.size());
+      result_tensors_data.push_back(tensors_data[indices_index]);
       ++indices_index;
     } else if (!tensors[i].CurrentTensorData()) {
       xla::ComputationClient::DataPtr xla_data = tensors[i].CurrentXlaData();
```

Reply:
This is neat!